xcm/xcm-builder/src/barriers.rs

```diff
@@ -93,26 +106,150 @@ impl<T: Contains<MultiLocation>> ShouldExecute for AllowTopLevelPaidExecutionFro
+			target: "xcm::barriers",
+			"WithComputedOrigin origin: {:?}, instructions: {:?}, max_weight: {:?}, weight_credit: {:?}",
+			origin, instructions, max_weight, weight_credit,
+		);
+		let mut actual_origin = *origin;
+		let mut skipped = 0;
+		// NOTE: We do not check the validity of `UniversalOrigin` here, meaning that a malicious
+		// origin could place a `UniversalOrigin` in order to spoof some location which gets free
+		// execution. This technical could get it past the barrier condition, but the execution
+		// would instantly fail since the first instruction would cause an error with the
+		// invalid UniversalOrigin.
+		while skipped < MaxPrefixes::get() as usize {
+			match instructions.get(skipped) {
+				Some(UniversalOrigin(new_global)) => {
+					// Note the origin is *relative to local consensus*! So we need to escape local
+					// consensus with the `parents` before diving in into the `universal_location`.
+					actual_origin = X1(*new_global).relative_to(&LocalUniversal::get());
+				},
+				Some(DescendOrigin(j)) => {
+					actual_origin.append_with(*j).map_err(|_| ())?;
+				},
+				_ => break,
+			}
+			skipped += 1;
+		}
+		InnerBarrier::should_execute(
+			&actual_origin,
+			&mut instructions[skipped..],
+			max_weight,
+			weight_credit,
+		)
+	}
+}
+
+/// Allows execution from any origin that is contained in `T` (i.e. `T::Contains(origin)`).
+///
+/// Use only for executions from completely trusted origins, from which no unpermissioned messages
+/// can be sent.
 pub struct AllowUnpaidExecutionFrom<T>(PhantomData<T>);
 impl<T: Contains<MultiLocation>> ShouldExecute for AllowUnpaidExecutionFrom<T> {
 	fn should_execute<RuntimeCall>(
 		origin: &MultiLocation,
-		_message: &mut Xcm<RuntimeCall>,
+		instructions: &mut [Instruction<RuntimeCall>],
 		_max_weight: Weight,
 		_weight_credit: &mut Weight,
 	) -> Result<(), ()> {
 		log::trace!(
 			target: "xcm::barriers",
-			"AllowUnpaidExecutionFrom origin: {:?}, message: {:?}, max_weight: {:?}, weight_credit: {:?}",
-			origin, _message, _max_weight, _weight_credit,
+			"AllowUnpaidExecutionFrom origin: {:?}, instructions: {:?}, max_weight: {:?}, weight_credit: {:?}",
+			origin, instructions, _max_weight, _weight_credit,
 		);
 		ensure!(T::contains(origin), ());
 		Ok(())
 	}
 }
 
+/// Allows execution from any origin that is contained in `T` (i.e. `T::Contains(origin)`) if the
+/// message begins with the instruction `UnpaidExecution`.
+///
+/// Use only for executions from trusted origin groups.
+pub struct AllowExplicitUnpaidExecutionFrom<T>(PhantomData<T>);
+impl<T: Contains<MultiLocation>> ShouldExecute for AllowExplicitUnpaidExecutionFrom<T> {
+	fn should_execute<Call>(
+		origin: &MultiLocation,
+		instructions: &mut [Instruction<Call>],
+		max_weight: Weight,
+		_weight_credit: &mut Weight,
+	) -> Result<(), ()> {
+		log::trace!(
+			target: "xcm::barriers",
+			"AllowUnpaidExecutionFrom origin: {:?}, instructions: {:?}, max_weight: {:?}, weight_credit: {:?}",
```

Comment:
@KiChjang 
just a very nit in log:
```suggestion
			"AllowExplicitUnpaidExecutionFrom origin: {:?}, instructions: {:?}, max_weight: {:?}, weight_credit: {:?}",
```